File: cpp/prover.cpp
/*
 * Proof Checker - C++ Implementation
 *
 * A native C++ theorem prover using Z3 SMT solver.
 * Mirrors the Python implementation with JSON input/output.
 *
 * Build: mkdir build && cd build && cmake .. && make
 * Usage: ./prover < input.json
 */

#include <iostream>
#include <map>
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <string>
#include <z3++.h>

using json = nlohmann::json;
using namespace z3;

class ProofError : public std::runtime_error {
public:
  explicit ProofError(const std::string &msg) : std::runtime_error(msg) {}
};

class TermError : public ProofError {
public:
  explicit TermError(const std::string &msg)
      : ProofError("Term error: " + msg) {}
};

class FormulaError : public ProofError {
public:
  explicit FormulaError(const std::string &msg)
      : ProofError("Formula error: " + msg) {}
};

// Use std::map with expr stored via optional wrapper
struct ExprWrapper {
  std::optional<expr> e;
  ExprWrapper() = default;
  ExprWrapper(expr ex) : e(std::move(ex)) {}
  operator expr() const { return *e; }
};

using Environment = std::map<std::string, ExprWrapper>;
using VarTypes = std::map<std::string, std::string>;

// Forward declarations
expr term_to_z3(const json &t, context &ctx, Environment &env,
                const VarTypes &var_types);
expr formula_to_z3(const json &f, context &ctx, Environment &env,
                   const VarTypes &var_types);

/**
 * Get or create a variable in the environment.
 * Uses var_types to determine whether to create Int or Real.
 */
expr get_var(const std::string &name, context &ctx, Environment &env,
             const VarTypes &var_types) {
  auto it = env.find(name);
  if (it != env.end()) {
    return it->second;
  }
  expr v = (var_types.count(name) && var_types.at(name) == "Int")
               ? ctx.int_const(name.c_str())
               : ctx.real_const(name.c_str());
  env[name] = ExprWrapper(v);
  return v;
}

/**
 * Convert a term JSON object to a Z3 expression.
 */
expr term_to_z3(const json &t, context &ctx, Environment &env,
                const VarTypes &var_types) {
  if (!t.is_object()) {
    throw TermError("Term must be an object");
  }

  if (!t.contains("type")) {
    throw TermError("Term missing 'type' field");
  }

  std::string ty = t["type"];

  if (ty == "num") {
    if (!t.contains("value")) {
      throw TermError("Numeric term missing 'value' field");
    }
    std::string val;
    if (t["value"].is_string()) {
      val = t["value"].get<std::string>();
    } else if (t["value"].is_number_integer()) {
      val = std::to_string(t["value"].get<int64_t>());
    } else {
      val = std::to_string(t["value"].get<double>());
    }
    return ctx.real_val(val.c_str());
  }

  if (ty == "var") {
    if (!t.contains("name")) {
      throw TermError("Variable term missing 'name' field");
    }
    return get_var(t["name"].get<std::string>(), ctx, env, var_types);
  }

  if (ty == "bin") {
    if (!t.contains("op") || !t.contains("lhs") || !t.contains("rhs")) {
      throw TermError("Binary term missing 'op', 'lhs', or 'rhs' field");
    }

    std::string op = t["op"];
    expr a = term_to_z3(t["lhs"], ctx, env, var_types);
    expr b = term_to_z3(t["rhs"], ctx, env, var_types);

    if (op == "+")
      return a + b;
    if (op == "-")
      return a - b;
    if (op == "*")
      return a * b;
    if (op == "/")
      return a / b;
    throw TermError("Unknown binary operator: " + op);
  }

  if (ty == "abs") {
    if (!t.contains("arg")) {
      throw TermError("Abs term missing 'arg' field");
    }
    expr x = term_to_z3(t["arg"], ctx, env, var_types);
    return ite(x >= 0, x, -x);
  }

  if (ty == "neg") {
    if (!t.contains("arg")) {
      throw TermError("Neg term missing 'arg' field");
    }
    expr x = term_to_z3(t["arg"], ctx, env, var_types);
    return -x;
  }

  if (ty == "pow") {
    if (!t.contains("base") || !t.contains("exp")) {
      throw TermError("Pow term missing 'base' or 'exp' field");
    }
    expr base = term_to_z3(t["base"], ctx, env, var_types);
    expr exp_val = term_to_z3(t["exp"], ctx, env, var_types);
    return pw(base, exp_val);
  }

  if (ty == "sqrt") {
    if (!t.contains("arg")) {
      throw TermError("Sqrt term missing 'arg' field");
    }
    expr x = term_to_z3(t["arg"], ctx, env, var_types);
    return pw(x, ctx.real_val("1/2"));
  }

  if (ty == "min") {
    if (!t.contains("args") || t["args"].size() < 2) {
      throw TermError("Min requires at least 2 arguments");
    }

    expr result = term_to_z3(t["args"][0], ctx, env, var_types);
    for (size_t i = 1; i < t["args"].size(); ++i) {
      expr other = term_to_z3(t["args"][i], ctx, env, var_types);
      result = ite(result <= other, result, other);
    }
    return result;
  }

  if (ty == "max") {
    if (!t.contains("args") || t["args"].size() < 2) {
      throw TermError("Max requires at least 2 arguments");
    }

    expr result = term_to_z3(t["args"][0], ctx, env, var_types);
    for (size_t i = 1; i < t["args"].size(); ++i) {
      expr other = term_to_z3(t["args"][i], ctx, env, var_types);
      result = ite(result >= other, result, other);
    }
    return result;
  }

  throw TermError("Unknown term type: " + ty);
}

/**
 * Convert a formula JSON object to a Z3 expression.
 */
expr formula_to_z3(const json &f, context &ctx, Environment &env,
                   const VarTypes &var_types) {
  if (!f.is_object()) {
    throw FormulaError("Formula must be an object");
  }

  if (!f.contains("type")) {
    throw FormulaError("Formula missing 'type' field");
  }

  std::string ty = f["type"];

  if (ty == "rel") {
    if (!f.contains("op") || !f.contains("lhs") || !f.contains("rhs")) {
      throw FormulaError(
          "Relational formula missing 'op', 'lhs', or 'rhs' field");
    }

    std::string op = f["op"];
    expr lhs = term_to_z3(f["lhs"], ctx, env, var_types);
    expr rhs = term_to_z3(f["rhs"], ctx, env, var_types);

    if (op == "<")
      return lhs < rhs;
    if (op == "<=")
      return lhs <= rhs;
    if (op == "=")
      return lhs == rhs;
    if (op == "!=")
      return lhs != rhs;
    if (op == ">")
      return lhs > rhs;
    if (op == ">=")
      return lhs >= rhs;
    throw FormulaError("Unknown relational operator: " + op);
  }

  if (ty == "and") {
    if (!f.contains("args")) {
      throw FormulaError("And formula missing 'args' field");
    }

    if (f["args"].empty()) {
      return ctx.bool_val(true); // Empty conjunction is true
    }

    expr_vector args(ctx);
    for (const auto &arg : f["args"]) {
      args.push_back(formula_to_z3(arg, ctx, env, var_types));
    }
    return mk_and(args);
  }

  if (ty == "or") {
    if (!f.contains("args")) {
      throw FormulaError("Or formula missing 'args' field");
    }

    if (f["args"].empty()) {
      return ctx.bool_val(false); // Empty disjunction is false
    }

    expr_vector args(ctx);
    for (const auto &arg : f["args"]) {
      args.push_back(formula_to_z3(arg, ctx, env, var_types));
    }
    return mk_or(args);
  }

  if (ty == "not") {
    if (!f.contains("arg")) {
      throw FormulaError("Not formula missing 'arg' field");
    }
    return !formula_to_z3(f["arg"], ctx, env, var_types);
  }

  if (ty == "implies") {
    if (!f.contains("lhs") || !f.contains("rhs")) {
      throw FormulaError("Implies formula missing 'lhs' or 'rhs' field");
    }
    expr lhs = formula_to_z3(f["lhs"], ctx, env, var_types);
    expr rhs = formula_to_z3(f["rhs"], ctx, env, var_types);
    return implies(lhs, rhs);
  }

  if (ty == "forall") {
    if (!f.contains("vars") || !f.contains("body")) {
      throw FormulaError("Forall formula missing 'vars' or 'body' field");
    }

    expr_vector quant_vars(ctx);
    for (const auto &name_json : f["vars"]) {
      std::string name = name_json.get<std::string>();
      expr v = (var_types.count(name) && var_types.at(name) == "Int")
                   ? ctx.int_const(name.c_str())
                   : ctx.real_const(name.c_str());
      env[name] = ExprWrapper(v);
      quant_vars.push_back(v);
    }

    expr body = formula_to_z3(f["body"], ctx, env, var_types);
    return forall(quant_vars, body);
  }

  if (ty == "exists") {
    if (!f.contains("vars") || !f.contains("body")) {
      throw FormulaError("Exists formula missing 'vars' or 'body' field");
    }

    expr_vector quant_vars(ctx);
    for (const auto &name_json : f["vars"]) {
      std::string name = name_json.get<std::string>();
      expr v = (var_types.count(name) && var_types.at(name) == "Int")
                   ? ctx.int_const(name.c_str())
                   : ctx.real_const(name.c_str());
      env[name] = ExprWrapper(v);
      quant_vars.push_back(v);
    }

    expr body = formula_to_z3(f["body"], ctx, env, var_types);
    return exists(quant_vars, body);
  }

  throw FormulaError("Unknown formula type: " + ty);
}

/**
 * Format a counterexample model.
 */
json format_model(const model &m, const Environment &env) {
  json model_out = json::object();
  for (const auto &[name, wrapper] : env) {
    expr var = wrapper;
    model_out[name] = m.eval(var, true).to_string();
  }
  return model_out;
}

/**
 * Main proof function.
 */
json prove(const json &req) {
  try {
    context ctx;
    Environment env;
    VarTypes var_types;

    // Parse variable types
    if (req.contains("var_types") && req["var_types"].is_object()) {
      for (auto &[name, type_val] : req["var_types"].items()) {
        var_types[name] = type_val.get<std::string>();
      }
    }

    // Initialize declared variables
    if (req.contains("vars")) {
      for (const auto &name_json : req["vars"]) {
        std::string n = name_json.get<std::string>();
        if (var_types.count(n) && var_types[n] == "Int") {
          env[n] = ExprWrapper(ctx.int_const(n.c_str()));
        } else {
          env[n] = ExprWrapper(ctx.real_const(n.c_str()));
        }
      }
    }

    solver s(ctx);

    // Add assumptions
    if (req.contains("assumptions")) {
      for (const auto &a : req["assumptions"]) {
        s.add(formula_to_z3(a, ctx, env, var_types));
      }
    }

    // Get the claim
    if (!req.contains("claim")) {
      return {{"ok", false},
              {"status", "error"},
              {"error", "Missing 'claim' field"}};
    }

    expr claim = formula_to_z3(req["claim"], ctx, env, var_types);

    // Prove: assumptions => claim
    // Check UNSAT of: assumptions AND (NOT claim)
    s.add(!claim);

    check_result result = s.check();

    if (result == unsat) {
      return {{"ok", true}, {"status", "proven"}};
    }

    if (result == sat) {
      model m = s.get_model();
      return {{"ok", false},
              {"status", "disproven"},
              {"model", format_model(m, env)}};
    }

    // result == unknown
    return {{"ok", false},
            {"status", "unknown"},
            {"message", "Z3 could not determine satisfiability"}};

  } catch (const ProofError &e) {
    return {{"ok", false}, {"status", "error"}, {"error", e.what()}};
  } catch (const z3::exception &e) {
    return {{"ok", false},
            {"status", "error"},
            {"error", std::string("Z3 error: ") + e.msg()}};
  } catch (const std::exception &e) {
    return {{"ok", false},
            {"status", "error"},
            {"error", std::string("Internal error: ") + e.what()}};
  }
}

int main() {
  try {
    json req;
    std::cin >> req;

    json result = prove(req);
    std::cout << result << std::endl;

    return result["ok"] ? 0 : 1;

  } catch (const json::parse_error &e) {
    json error = {{"ok", false},
                  {"status", "error"},
                  {"error", std::string("Invalid JSON: ") + e.what()}};
    std::cout << error << std::endl;
    return 1;
  }
}

File: python/prover.py
"""
Proof Checker - Extended Prover
Supports logical connectives, quantifiers, and rich arithmetic.
"""

import json
import sys
from z3 import (
    Real, RealVal, Int, IntVal, If, Solver, Not, And, Or, Implies, ForAll, Exists,
    sat, unsat, unknown
)


class ProofError(Exception):
    """Base exception for proof errors."""
    pass


class TermError(ProofError):
    """Error in term parsing."""
    pass


class FormulaError(ProofError):
    """Error in formula parsing."""
    pass


def term_to_z3(t, env, var_types=None):
    """Convert a term JSON object to a Z3 expression.
    
    Supported term types:
    - num: numeric literal
    - var: variable reference
    - bin: binary operation (+, -, *, /)
    - abs: absolute value
    - neg: unary negation
    - pow: exponentiation
    - sqrt: square root
    - min: minimum of arguments
    - max: maximum of arguments
    
    Args:
        t: Term JSON object
        env: Variable environment mapping names to Z3 expressions
        var_types: Optional dict mapping variable names to types ('Int' or 'Real')
    """
    if var_types is None:
        var_types = {}
        
    if not isinstance(t, dict):
        raise TermError(f"Term must be an object, got: {type(t).__name__}")
    
    ty = t.get("type")
    if ty is None:
        raise TermError(f"Term missing 'type' field: {t}")

    if ty == "num":
        value = t.get("value")
        if value is None:
            raise TermError("Numeric term missing 'value' field")
        return RealVal(value)

    if ty == "var":
        name = t.get("name")
        if name is None:
            raise TermError("Variable term missing 'name' field")
        if name not in env:
            # Create Int or Real based on var_types
            if var_types.get(name) == 'Int':
                env[name] = Int(name)
            else:
                env[name] = Real(name)
        return env[name]

    if ty == "bin":
        op = t.get("op")
        if op is None:
            raise TermError("Binary term missing 'op' field")
        if "lhs" not in t or "rhs" not in t:
            raise TermError("Binary term missing 'lhs' or 'rhs' field")
        
        a = term_to_z3(t["lhs"], env, var_types)
        b = term_to_z3(t["rhs"], env, var_types)
        
        if op == "+": return a + b
        if op == "-": return a - b
        if op == "*": return a * b
        if op == "/": return a / b
        raise TermError(f"Unknown binary operator: {op}")

    if ty == "abs":
        if "arg" not in t:
            raise TermError("Abs term missing 'arg' field")
        x = term_to_z3(t["arg"], env, var_types)
        return If(x >= 0, x, -x)

    if ty == "neg":
        if "arg" not in t:
            raise TermError("Neg term missing 'arg' field")
        x = term_to_z3(t["arg"], env, var_types)
        return -x

    if ty == "pow":
        if "base" not in t or "exp" not in t:
            raise TermError("Pow term missing 'base' or 'exp' field")
        base = term_to_z3(t["base"], env, var_types)
        exp = term_to_z3(t["exp"], env, var_types)
        return base ** exp

    if ty == "sqrt":
        if "arg" not in t:
            raise TermError("Sqrt term missing 'arg' field")
        x = term_to_z3(t["arg"], env, var_types)
        # sqrt(x) represented as x^0.5
        return x ** RealVal("0.5")

    if ty == "min":
        args = t.get("args", [])
        if len(args) < 2:
            raise TermError("Min requires at least 2 arguments")
        result = term_to_z3(args[0], env, var_types)
        for arg in args[1:]:
            other = term_to_z3(arg, env, var_types)
            result = If(result <= other, result, other)
        return result

    if ty == "max":
        args = t.get("args", [])
        if len(args) < 2:
            raise TermError("Max requires at least 2 arguments")
        result = term_to_z3(args[0], env, var_types)
        for arg in args[1:]:
            other = term_to_z3(arg, env, var_types)
            result = If(result >= other, result, other)
        return result

    raise TermError(f"Unknown term type: {ty}")


def formula_to_z3(f, env, var_types=None):
    """Convert a formula JSON object to a Z3 expression.
    
    Supported formula types:
    - rel: relational comparison (<, <=, =, !=, >, >=)
    - and: conjunction of formulas
    - or: disjunction of formulas
    - not: negation of a formula
    - implies: implication (lhs => rhs)
    - forall: universal quantifier
    - exists: existential quantifier
    """
    if var_types is None:
        var_types = {}
        
    if not isinstance(f, dict):
        raise FormulaError(f"Formula must be an object, got: {type(f).__name__}")
    
    ty = f.get("type")
    if ty is None:
        raise FormulaError(f"Formula missing 'type' field: {f}")

    if ty == "rel":
        op = f.get("op")
        if op is None:
            raise FormulaError("Relational formula missing 'op' field")
        if "lhs" not in f or "rhs" not in f:
            raise FormulaError("Relational formula missing 'lhs' or 'rhs' field")
        
        lhs = term_to_z3(f["lhs"], env, var_types)
        rhs = term_to_z3(f["rhs"], env, var_types)

        if op == "<":  return lhs < rhs
        if op == "<=": return lhs <= rhs
        if op == "=":  return lhs == rhs
        if op == "!=": return lhs != rhs
        if op == ">":  return lhs > rhs
        if op == ">=": return lhs >= rhs
        raise FormulaError(f"Unknown relational operator: {op}")

    if ty == "and":
        args = f.get("args", [])
        if len(args) == 0:
            return True  # Empty conjunction is true
        z3_args = [formula_to_z3(arg, env, var_types) for arg in args]
        return And(*z3_args)

    if ty == "or":
        args = f.get("args", [])
        if len(args) == 0:
            return False  # Empty disjunction is false
        z3_args = [formula_to_z3(arg, env, var_types) for arg in args]
        return Or(*z3_args)

    if ty == "not":
        if "arg" not in f:
            raise FormulaError("Not formula missing 'arg' field")
        inner = formula_to_z3(f["arg"], env, var_types)
        return Not(inner)

    if ty == "implies":
        if "lhs" not in f or "rhs" not in f:
            raise FormulaError("Implies formula missing 'lhs' or 'rhs' field")
        lhs = formula_to_z3(f["lhs"], env, var_types)
        rhs = formula_to_z3(f["rhs"], env, var_types)
        return Implies(lhs, rhs)

    if ty == "forall":
        var_names = f.get("vars", [])
        if not var_names:
            raise FormulaError("Forall formula missing 'vars' field")
        if "body" not in f:
            raise FormulaError("Forall formula missing 'body' field")
        
        # Create fresh variables for the quantifier
        quant_vars = []
        for name in var_names:
            if var_types.get(name) == 'Int':
                v = Int(name)
            else:
                v = Real(name)
            env[name] = v
            quant_vars.append(v)
        
        body = formula_to_z3(f["body"], env, var_types)
        return ForAll(quant_vars, body)

    if ty == "exists":
        var_names = f.get("vars", [])
        if not var_names:
            raise FormulaError("Exists formula missing 'vars' field")
        if "body" not in f:
            raise FormulaError("Exists formula missing 'body' field")
        
        # Create fresh variables for the quantifier
        quant_vars = []
        for name in var_names:
            if var_types.get(name) == 'Int':
                v = Int(name)
            else:
                v = Real(name)
            env[name] = v
            quant_vars.append(v)
        
        body = formula_to_z3(f["body"], env, var_types)
        return Exists(quant_vars, body)

    raise FormulaError(f"Unknown formula type: {ty}")


def format_counterexample(model, env):
    """Format a Z3 model as a human-readable counterexample."""
    lines = ["Counterexample found:"]
    for name, var in sorted(env.items()):
        value = model.eval(var, model_completion=True)
        lines.append(f"  {name} = {value}")
    return "\n".join(lines)


def prove(assumptions, claim, declared_vars=None, var_types=None, steps=None):
    """Attempt to prove that assumptions imply the claim.
    
    Args:
        assumptions: List of assumption formulas
        claim: The claim to prove
        declared_vars: Optional list of declared variable names
        var_types: Optional dict mapping variable names to types ('Int' or 'Real')
        steps: Optional list of intermediate proof steps to verify
    
    Returns a dict with:
    - ok: True if proven, False if counterexample found
    - model: counterexample variable assignments (if ok=False)
    - error: error message (if parsing failed)
    - status: "proven", "disproven", "unknown", or "error"
    - step_results: results for each intermediate step (if steps provided)
    """
    try:
        if var_types is None:
            var_types = {}
        if steps is None:
            steps = []
            
        env = {}
        if declared_vars:
            for name in declared_vars:
                if var_types.get(name) == 'Int':
                    env[name] = Int(name)
                else:
                    env[name] = Real(name)

        # Verify intermediate steps first
        step_results = []
        current_assumptions = list(assumptions)
        
        for i, step in enumerate(steps):
            # Handle cases step type
            if step.get("type") == "cases":
                cases = step.get("cases", [])
                case_results = []
                all_cases_ok = True
                
                # Collect all case conditions for exhaustiveness check
                conditions = []
                
                for case_idx, case in enumerate(cases):
                    condition = case.get("condition")
                    conditions.append(condition)
                    case_steps = case.get("steps", [])
                    
                    # For each case, assume the condition and verify steps
                    case_assumptions = current_assumptions + [condition]
                    for cs in case_steps:
                        cs_formula = cs.get("formula")
                        if cs_formula:
                            # Check if this step is provable under the case assumption
                            s = Solver()
                            for a in case_assumptions:
                                s.add(formula_to_z3(a, env.copy(), var_types))
                            s.add(Not(formula_to_z3(cs_formula, env.copy(), var_types)))
                            if s.check() == unsat:
                                case_assumptions.append(cs_formula)
                    
                    case_results.append({"case": case_idx + 1, "ok": True})
                
                # Check if cases are exhaustive: disjunction of conditions should be tautology
                # (given current assumptions)
                if conditions:
                    s = Solver()
                    for a in current_assumptions:
                        s.add(formula_to_z3(a, env.copy(), var_types))
                    
                    # Negate (cond1 OR cond2 OR ...) 
                    cond_z3 = [formula_to_z3(c, env.copy(), var_types) for c in conditions]
                    s.add(Not(Or(*cond_z3)))
                    
                    exhaustive = s.check() == unsat
                    if not exhaustive:
                        all_cases_ok = False
                        step_results.append({
                            "step": i + 1,
                            "type": "cases",
                            "ok": False,
                            "status": "non-exhaustive",
                            "message": "Cases may not cover all possibilities"
                        })
                    else:
                        step_results.append({
                            "step": i + 1,
                            "type": "cases",
                            "ok": True,
                            "status": "proven",
                            "case_results": case_results
                        })
                continue
            
            step_formula = step.get("formula")
            if step_formula is None:
                step_results.append({"step": i + 1, "ok": False, "error": "Step missing formula"})
                continue
            
            s = Solver()
            for a in current_assumptions:
                s.add(formula_to_z3(a, env.copy(), var_types))
            
            step_z3 = formula_to_z3(step_formula, env.copy(), var_types)
            s.add(Not(step_z3))
            
            result = s.check()
            
            if result == unsat:
                step_results.append({"step": i + 1, "ok": True, "status": "proven"})
                # Add the proven step as an assumption for subsequent steps
                current_assumptions.append(step_formula)
            elif result == sat:
                m = s.model()
                model_out = {name: str(m.eval(v, model_completion=True)) for name, v in env.items()}
                step_results.append({
                    "step": i + 1, 
                    "ok": False, 
                    "status": "disproven",
                    "model": model_out
                })
                # Don't add unproven steps to assumptions
            else:
                step_results.append({"step": i + 1, "ok": False, "status": "unknown"})

        # Now prove the final claim
        s = Solver()
        
        # Add all original assumptions plus proven steps
        for a in current_assumptions:
            s.add(formula_to_z3(a, env, var_types))

        # Convert the claim
        claim_z3 = formula_to_z3(claim, env, var_types)

        # Prove: assumptions => claim
        # Check UNSAT of: assumptions AND (NOT claim)
        s.add(Not(claim_z3))

        result = s.check()
        
        if result == unsat:
            response = {"ok": True, "status": "proven"}
            if step_results:
                response["step_results"] = step_results
            return response
        
        if result == sat:
            m = s.model()
            model_out = {
                name: str(m.eval(v, model_completion=True)) 
                for name, v in env.items()
            }
            response = {
                "ok": False, 
                "status": "disproven",
                "model": model_out,
                "message": format_counterexample(m, env)
            }
            if step_results:
                response["step_results"] = step_results
            return response
        
        # result == unknown
        response = {
            "ok": False,
            "status": "unknown",
            "message": "Z3 could not determine satisfiability (timeout or incomplete theory)"
        }
        if step_results:
            response["step_results"] = step_results
        return response

    except (TermError, FormulaError) as e:
        return {
            "ok": False,
            "status": "error",
            "error": str(e)
        }
    except Exception as e:
        return {
            "ok": False,
            "status": "error",
            "error": f"Internal error: {e}"
        }


def main():
    """Main entry point - reads JSON from stdin, outputs JSON to stdout."""
    try:
        req = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        json.dump({"ok": False, "status": "error", "error": f"Invalid JSON: {e}"}, sys.stdout)
        return

    declared_vars = req.get("vars", [])
    var_types = req.get("var_types", {})
    assumptions = req.get("assumptions", [])
    steps = req.get("steps", [])
    claim = req.get("claim")
    
    if claim is None:
        json.dump({"ok": False, "status": "error", "error": "Missing 'claim' field"}, sys.stdout)
        return

    result = prove(assumptions, claim, declared_vars, var_types, steps)
    json.dump(result, sys.stdout)


if __name__ == "__main__":
    main()

